File: codeExamples/week11/spring24_insta_cult-main/client/src/services/CultsService.js
import { AppState } from "../AppState.js"
import { Cult } from "../models/Cult.js"
import { logger } from "../utils/Logger.js"
import { api } from "./AxiosService.js"

class CultsService {
  async createCult(cultData) {
    const res = await api.post('api/cults', cultData)
    logger.log('CREATED CULT 💀', res.data)
    // No need to push into AppState, since we can only create on a page that doesn't access any data from AppState
  }
  async destroyCult(cultId) {
    const res = await api.delete(`api/cults/${cultId}`)
    logger.log("DELETE CULT ❌", res.data)
    AppState.activeCult = null
  }
  async getCultById(cultId) {
    AppState.activeCult = null
    const res = await api.get(`api/cults/${cultId}`)
    logger.log('GOT CULT BY ID 💀', res.data)
    AppState.activeCult = new Cult(res.data)
  }
  async getCults() {
    AppState.cults.length = 0 // empties array
    // AppState.cults = []
    const res = await api.get('api/cults')
    logger.log('GOT CULTS 💀💀💀', res.data)
    AppState.cults = res.data.map(pojo => new Cult(pojo))
  }
}

export const cultsService = new CultsService()

File: codeExamples/week11/spring24_insta_cult-main/server/Services/CultsService.cs


namespace insta_cult.Services;

public class CultsService
{
  private readonly CultsRepository _repository;

  public CultsService(CultsRepository repository)
  {
    _repository = repository;
  }

  internal Cult CreateCult(Cult cultData)
  {
    Cult cult = _repository.CreateCult(cultData);
    return cult;
  }

  internal void DestroyCult(int cultId, string userId)
  {
    Cult cult = GetCultById(cultId);

    if (cult.LeaderId != userId)
    {
      throw new Exception("NOT YOUR CULT");
    }

    _repository.DestroyCult(cultId);
  }

  internal List<Cult> GetAllCults()
  {
    List<Cult> cults = _repository.GetAllCults();
    return cults;
  }

  internal Cult GetCultById(int cultId)
  {
    Cult cult = _repository.GetCultById(cultId);

    if (cult == null)
    {
      throw new Exception($"Invalid cult id: {cultId}");
    }

    return cult;
  }
}
